c_emulator/riscv_sim.cpp

```
@@ -40,62 +41,83 @@ enum {
 bool config_print_reg = true;
 bool config_print_mem_access = true;
 bool config_print_platform = true;
-bool config_use_abi_names = false;
 bool config_print_rvfi = false;
 bool config_print_step = false;
+
+bool config_use_abi_names = false;
 bool config_enable_rvfi = false;
 
-void set_config_print(char *var, bool val)
-{
-  if (var == NULL || strcmp("all", var) == 0) {
-    config_print_instr = val;
-    config_print_mem_access = val;
-    config_print_reg = val;
-    config_print_platform = val;
-    config_print_rvfi = val;
-  } else if (strcmp("instr", var) == 0) {
-    config_print_instr = val;
-  } else if (strcmp("reg", var) == 0) {
-    config_print_reg = val;
-  } else if (strcmp("abi", var) == 0) {
-    config_use_abi_names = val;
-  } else if (strcmp("mem", var) == 0) {
-    config_print_mem_access = val;
-  } else if (strcmp("rvfi", var) == 0) {
-    config_print_rvfi = val;
-  } else if (strcmp("platform", var) == 0) {
-    config_print_platform = val;
-  } else if (strcmp("step", var) == 0) {
-    config_print_step = val;
-  } else {
-    fprintf(stderr, "Unknown trace category: '%s' (should be %s)\n", var,
-            "instr|reg|mem|rvfi|platform|step|all");
-    exit(EXIT_FAILURE);
-  }
-}
+std::map<std::string, std::tuple<bool *, std::string>> printers {
+    {"instr",    {&config_print_instr, "Instruction execution"}     },
+    {"reg",      {&config_print_reg, "Register accesses"}           },
+    {"mem",      {&config_print_mem_access, "Memory accesses"}      },
+    {"rvfi",     {&config_print_rvfi, "RVFI tracing"}               },
+    {"platform",
+     {&config_print_platform, "Privilege changes, MMIO, interrupts"}},
+    {"step",     {&config_print_step, "Steps"}                      },
+    {"all",      {NULL, "All of the other tracers"}                 },
+};
+
+std::function<void(const std::string &)> enable_printer {
+    [](const std::string &opt) {
+      auto ent = printers.find(opt);
+      if (ent != printers.end()) {
```

Will this silently ignore incorrect arguments?

---

No, the second example I gave showed this case:
```
$ ./build/c_emulator/sail_riscv_sim --no-trace foo  build/test/2025-06-22/riscv-tests/rv64ui-p-addw 
--no-trace: foo not in {all,instr,mem,platform,reg,rvfi,step}
Run with --help for more information.
```
The `->check(CLI::IsMember(printers))` takes care of it.